### PR TITLE
VaultAuth GCP workload identity support

### PR DESCRIPTION
**Pull request by tvoran**
Adds VaultAuth support for Vault's GCP auth method when using GKE workload identity.

Draft docs PR on usage: https://github.com/hashicorp/vault/pull/23684

Adds an extra [`terraform-gcp` folder](https://github.com/hashicorp/vault-secrets-operator/pull/411/files#diff-4910083ba14951c00a5244b5f1dbb1505249ed839d5e355877b6c255fabea0b8) for the vaultauthmethods integration tests that sets up the GCP/GKE bits for testing workload identity. It's a separate folder since provider blocks in terraform don't support counts/conditionals. And it's not in the infra gke terraform since the workload identity relationship can only be set to a specific k8s test namespace (instead of a [namespace wildcard like we do with AWS IRSA](https://github.com/hashicorp/vault-secrets-operator/blob/8365c7c07d278ed50e8fb91b9d7719df18a96527/test/integration/infra/eks/aws_auth.tf#L11)).

Example integration test run on GCP:

```shell-session
$ make -f gcp.mk integration-test-gke TESTARGS="-count 1 -v -run TestVaultAuth" TEST_WITH_HELM=true
...
--- PASS: TestVaultAuthMethods (100.56s)
    --- PASS: TestVaultAuthMethods/vaultauth-test-kubernetes (8.34s)
    --- PASS: TestVaultAuthMethods/vaultauth-test-jwt-serviceaccount (1.11s)
    --- PASS: TestVaultAuthMethods/vaultauth-test-jwt-secret (0.10s)
    --- PASS: TestVaultAuthMethods/vaultauth-test-approle (0.10s)
    --- SKIP: TestVaultAuthMethods/vaultauth-test-aws-irsa (0.00s)
    --- SKIP: TestVaultAuthMethods/vaultauth-test-aws-node (0.00s)
    --- SKIP: TestVaultAuthMethods/vaultauth-test-aws-instance-profile (0.00s)
    --- SKIP: TestVaultAuthMethods/vaultauth-test-aws-static (0.00s)
    --- PASS: TestVaultAuthMethods/vaultauth-test-gcp-workload-identity (40.43s)
    --- PASS: TestVaultAuthMethods/vaultauth-test-gcp-workload-identity-all-options (1.29s)
PASS
ok  	github.com/hashicorp/vault-secrets-operator/test/integration	101.224s
```

### CODE REVIEW

File: test/integration/vaultauthmethods/terraform/main.tf
Comment (benashz) [inline test/integration/vaultauthmethods/terraform/main.tf:18]: I don't see where this is being used. Is it still needed?

File: api/v1beta1/vaultauth_types.go
Comment (tomhjp) [inline api/v1beta1/vaultauth_types.go:115]: nit: Not sure if this is consistently followed everywhere, but ID is the Go naming convention, and we do use it for RoleID above.

```suggestion
	ProjectID string `json:"projectId,omitempty"`
```

File: internal/credentials/vault/gcp.go
Comment (tomhjp) [inline internal/credentials/vault/gcp.go:145]: I don't think it's crucial as GKE will cover almost everyone, but do we know how much more work it would be to also support running from GCP VMs? (I'm assuming this URL format means we only support GKE currently)

File: internal/credentials/vault/gcp.go
Comment (tomhjp) [inline internal/credentials/vault/gcp.go:118]: nit:

```suggestion
	GKEClusterName string
```

**Review by tomhjp — APPROVED** [review]
LGTM modulo Ben's comments. Nice work!

**Review by benashz — COMMENTED** [review]
GitHub seems to think I still need to "add a review"...

File: api/v1beta1/vaultauth_types.go
Comment (tvoran) [reply api/v1beta1/vaultauth_types.go:102]: Maybe "in the same Kubernetes namespace as the Vault*Secret referencing this resource"?

File: test/integration/vaultauthmethods_integration_test.go
Comment (tvoran) [reply test/integration/vaultauthmethods_integration_test.go:84]: > Is there a reason we are not getting the GCP project as well?
Just that there's no default for the project. I went ahead and removed this block since setting a default region on this end didn't seem necessary.

File: test/integration/vaultauthmethods/terraform/main.tf
Comment (tvoran) [reply test/integration/vaultauthmethods/terraform/main.tf:18]: Nope, not needed anymore.

File: test/integration/vaultauthmethods/terraform-gcp/main.tf
Comment (tvoran) [reply test/integration/vaultauthmethods/terraform-gcp/main.tf:37]: Yep, added in 95d19b1

File: test/integration/vaultauthmethods_integration_test.go
Comment (tvoran) [reply test/integration/vaultauthmethods_integration_test.go]: Sure, changed to `backoff` in 95d19b1

File: test/integration/integration_test.go
Comment (tvoran) [reply test/integration/integration_test.go]: Yeah, it's "Google Service Account". I think I fixed most of the naming issues in 95d19b1.

File: internal/credentials/vault/gcp.go
Comment (tvoran) [reply internal/credentials/vault/gcp.go:145]: Yeah this only supports GKE's workload identity. We may want to support setting a static credential file or using the underlying compute credentials in GKE someday, but workload identity seemed like the best option to add first until there's sufficient demand for the others, since this is all running inside kubernetes.

For running on GCP VMs, would that just be vanilla kubernetes on GCP compute instances?

Comment (tomhjp) [reply internal/credentials/vault/gcp.go:145]: > For running on GCP VMs, would that just be vanilla kubernetes on GCP compute instances?

That's what I was thinking of, with some kind of more generic GCE workload identity, but I think I was imagining a feature that doesn't exist, and maybe getting mixed up with [federated workload identity](https://cloud.google.com/iam/docs/workload-identity-federation-with-kubernetes#terraform) which is more about authenticating to GCP from outside GCP.

**Comment by kschoche** [comment]
Looking good so far and I'll defer to the other reviewers for now, but wanted to check if we were planning on adding support for this to the defaultAuthMethod chart template and values.yaml (which sure is getting large with each additional supported auth method)?

File: internal/credentials/vault/gcp.go
Comment (tvoran) [reply internal/credentials/vault/gcp.go:185]: Yeah, I think so. Added checks in b153aba.

**Review by benashz — APPROVED** [review]
LGTM! Great work!

**Comment by tvoran** [comment]
Thanks!

**Pull request merged**
